### ISSUE TITLE
[FEAT] Create Draw Shape from Image/Mask, or from Transparent sections of Image on Map Layer

### ISSUE THREAD

**Issue by PalasX** [Kruptein/PlanarAlly#641]
**Is your feature request related to a problem? Please describe.**
The drawing tools to create shapes that block lights and movement, like walls and doors, are great, but can be time consuming and tedious for very large maps. Anything that takes a human a long time to do, and has to be done repeatedly, is prime territory for automation.

**Describe the solution you'd like**
After dragging an asset to the MAP layer, and resizing according to grid size using the MAP TOOL, provide a means to automatically generate a single solid shape from the transparent parts of the image. Alternatively, let the user upload an image containing transparency to be used as a mask to create a shape.

**Describe alternatives you've considered**
Import a standard library for edge detection on a map asset image to try and create a single or series of shapes for walls? It seems more complicated this way.

**Additional context**
Example Map:
![image](https://user-images.githubusercontent.com/37154966/105404733-4dc97f00-5bf8-11eb-9c47-f2aad7fce39e.png)

Example resulting shape (white being transparent), or secondary mask image asset to be used as a "everything that isnt transparent will block light/vision as per user settings":
![image](https://user-images.githubusercontent.com/37154966/105404895-823d3b00-5bf8-11eb-9c5c-901becb5e52d.png)

I realize this could be a huge undertaking, but hoped that since there was so much python to deal with images already, that sorting out which pixels were and werent transparent would at least be near-trivial. I assume translating a 1bit (black/transparent) image into a shape would be the hard part. It could save DMs a significant amount of time and effort when using randomly generation maps, and people that create their own map art assets could export a "pathing" layer of their image maybe?

Thanks again for all the amazing work you do!

**Comment by Kruptein** [comment]
Hey, this is indeed something that would immensely help DMs all around.

The complexity lies in how the lighting/vision actually is calculated. It's based on an algorithm that uses line segments to construct a triangulation that can be quickly queried.

So for all existing shapes when toggling the 'blocks vision' option, it adds all the lines of the shape to the triangulation and the lines are known because they are either simple shapes or hand-drawn polygons.

I'm sure that I should be able to get the lines from a more complex shape such as your mask, so I'm sure it's feasible, it will just as you mentioned maybe take some more time.

Might look into this in the near future! There are some other things that I would like to tackle first, but this one would be nice to have.

Also nice that you provided an example with a mask ;)

**Comment by PalasX** [comment]
Continuing with the donjon exaple, would it be easier to directly import a .TSV file for a map, step through the TabSeperatedValue format, and build vector line segments from scratch? There is a python project donjon-painter on github that takes TSV and outputs PNG by building a tile based raster, but it might be a good start to grab the parsing for the TSV format?

Alternatively, what about building vector lines from an SVG? User could pre-convert their raster image to SVG with a tool like ImageMagik or something. Obviously more complex raster images would result in an SVG thats just too big to the point of unusability, but might be doable on a 1bit image.

And yeah, i realize this is a "wouldn't it be nice" feature for down the road, and would be a large undertaking. Slow and steady, im sure none of us wants to subject you to burnout :)

**Comment by Kruptein** [comment]
the SVG approach is something a friend of mine has also suggested in the past and should be rather easy to get the lines from indeed.

Also just as a small clarification, most of the heavy lifting is done in javascript in the browser. The python backend is primarily a dumb server that handles authentication, authorization, persistence and message passing.  Although nothing prevents me to implement some heavier things there if the ecosystem is better which might be the case for image processing.

**Comment by Kruptein** [comment]
Got some good news here, a friend of mine did some experiments way back in may related to svg stuff. I've looked at it and played around with some things and it's actually pretty promising.

I uploaded your example mask above to vectorizer to convert it to a svg and used some debug draw tools I have together with my friend's code and got this:

![image](https://user-images.githubusercontent.com/1814713/107855336-e59a4180-6e21-11eb-8f01-4100e2067bc2.png)

Ignore that one line in the bottom left, that's my draw tool connecting the last and first point, which is unnecessary here.

Next up is figuring out some scaling things, because we'll want to scale the fow lines along with the shape itself.

**Comment by Kruptein** [comment]
Expecting to actually have some working prototype tomorrow

**Comment by PalasX** [comment]
That is SO WICKED COOL! Thanks to you and your friend for finding an awesome way to implement this!!!

**Comment by PalasX** [comment]
I must be doing something wrong, i cannot get the SVG import to work.

I place a map image on the map layer, select it, view the properties, and select upload SVG. Then i select the walls SVG from the asset manager, aaaaaaaand nothing happens.

Do i need to do something special when crafting the SVG?
[map_101x101.svg.txt](https://github.com/Kruptein/PlanarAlly/files/6261474/map_101x101.svg.txt)
![hate3](https://user-images.githubusercontent.com/37154966/113647276-87871f00-9658-11eb-925b-9551552f4bfb.png)

**Comment by PalasX** [comment]
aaaaah, i had to turn on "Blocks light" and "Blocks movement" for the properties of the map asset itself on the map layer. is that how it's supposed to work?

**Comment by Kruptein** [comment]
Yeah, I suppose it shoud automatically toggle those on when you add an svg

**Comment by PalasX** [comment]
That's probably the best behavior, yeah. The process works beautifully,. by the way. Just gotta find a better offline vectorizer. Thanks again, your continued work on this project keeps making the world of difference to our gaming group!

**Comment by PalasX** [comment]
SVGs created with vectorizer.io work, but those that are outputed by dungeonscrawler (even after removing ALL the extra thatching and fluff and background so that you only have a 2 color, unfilled, outline of the map) do not create any walls.

The SVG from vectorizer.io seems to be using PATH elements, whereas the ones coming out of use POLYGON. Is this what's causing the flub?

vectorizer -
<g fill="none" stroke="#000000" stroke-width="2.5">
  <path d="M2050.3 4974.3 l0 -25.3 99.7 0 99.7 0 0 -24 .............. etc etc etc

dungeonscrawler, https://probabletrain.itch.io/dungeon-scrawl, all background, thatching, shadows, etc disabled -
<rect width="1920" height="947" stroke-linejoin="miter" stroke-linecap="butt" fill="#ffffff" fill-opacity="1" stroke="none" stroke-width="0.1469332310991199"/><g><g mask="url(&quot;#SvgjsMask1002&quot;)"><polygon points="477.6595044494085,56.924455795979505 599.3202932660953,56.924455795979505 599.3202932660953,62.21412558216337 599.3202932660953,62.21412558216337 599.3202932660953,67.50364843511613 599.320 .........etc etc etc

**Comment by dthv** [comment]
@PalasX just in case there will be some sort of FAQ and/or list of tricks or something – may I copy your comment and/or parts thereof into the documentation?

**Comment by PalasX** [comment]
@dthv I consent entirely to the free use, without any further stipulations or addendums, of my previous post, in whatever capacity anyone wants to use. Go for it, i wrote it all down 50% to get it straight in my head and document what i was doing so i wouldnt forget, and 50% to help anyone else.

Additionally, mirroring the image vertically MAY NOT be necessary for some maps. I had to do it for all "normal" maps so far, but not when creating "cavern" type maps. Also, do NOT try to do this with large cavern type maps. Everything will work fine until you turn on enable blocking light/movement, then it will all grind to a halt. On the plus side, cavern player maps from donjon dont have doors or anything, so you can use the player map right away without feeding the tsv through dungeon crawler (which would also make it blocky again, and not cavern-y).

Also additionally, you could probably do the conversion from PNG to 2bit BMP in the terminal/command line with ImageMagick, and thus skip the MS Paint requirement. It's what i'll end up doing in the future.

tl;dr: Yes. You can use my post, the contents therein, or any part of it, in any manner you choose.

**Comment by PalasX** [comment]
possible solution to relying on an online tool for TSV to SVG: https://thetrove.is/Resources/Programs/Donjon_TSV_to_SVG_converter.zip

Outputs an svg from a tsv, but makes a separate <path> for each tile, and does not output some header into like doctype and viewport sizes, but those can be added manually with any text editor. released as a compile .JAR, with no source available :(

**Comment by PalasX** [comment]
Using ImageMagick and the above linked TSV_to_SVG converter, we can avoid the use of ANY online tools:

TSVtoSVG, rename TSV to map.txt, run DonjonTSVtoSVGconverter.jar, get mapoutYxZ.txt

rename mapoutYxZ.txt mapoutYxZ.svg

ImageMagick> convert mapoutYxZ.svg -resize 8050x8050 -background white -gravity center -extent 8050x8050 mapoutYxZ.bmp
where 8050 is the size in pixels of the player map, minus 1.

ImageMagick> convert mapoutYxZ.bmp -flip mapoutYxZ.bmp

potrace.exe -u 1.333333 -x 0.1 -s -O 1 mapoutYxZ.bmp

upload to planar along with the player map, or whatever stylized version you want. match the map up to your grid, then apply the svg.


This process skips the need to monochrome the BMP first, but DOES NOT WORK ON CAVE STYLE MAPS! Well, it will, but it will reduce them to the squared off style and not keep the cave-y-ness of them.

**Comment by Kruptein** [comment]
Hey there quite a bunch of chat I missed :D

Could someone summarize if there is something that needs to happen / be checked on my end?
I see some things related to flipped images for example.

**Comment by PalasX** [comment]
Just wanted to update that this step is no longer needed since the SVG handling has been updated to account for transforms:
ImageMagick> convert mapoutYxZ.bmp -flip mapoutYxZ.bmp

verified that omitting it from the process still yields a perfect working SVG from a donjon TSV. My continued appreciation for the project and respect to it's contributors <3
